### PR TITLE
s/cgroup, systemd: escape systemd unit names in CreateTransientScopeForTracking

### PR DESCRIPTION
**Pull request by andrewphelpsj**
With the addition of component hooks, we'll have unit names that include a `+`, like `snap.snapname+comp.hook.install`. This causes systemd to complain that the unit isn't properly escaped. On the command line, systemd-run will properly escape this for you (with a warning), but the dbus API doesn't do that.

Output from failure:
```
logger.go:93: DEBUG: -- snap startup {"stage":"start", "time":"1711562759.579290"}
logger.go:93: DEBUG: executing snap-confine from /snap/snapd/x1/usr/lib/snapd/snap-confine
logger.go:93: DEBUG: SELinux not enabled
logger.go:93: DEBUG: creating transient scope snap.snap-with-comps+comp1.hook.install
logger.go:93: DEBUG: StartTransientUnit failed with "org.freedesktop.DBus.Error.InvalidArgs": [Invalid unit name or type.]
error: cannot create transient scope: DBus error "org.freedesktop.DBus.Error.InvalidArgs": [Invalid unit name or type.]
```

### CODE REVIEW

File: systemd/escape.go [hunk 1]
```diff
@@ -61,9 +71,21 @@ func EscapeUnitNamePath(in string) string {
...
+			fmt.Fprintf(&buf, `\x%x`, []byte{in[i]})
 		}
 	}
 
 	return buf.String()
 }
+
+// EscapeUnitNamePath works like systemd-escape --path
+// FIXME: we could use github.com/coreos/go-systemd/unit/escape.go and EscapePath
+// from it. But that's not in the archive and it won't work with go1.3
```
Comment (bboozzoo) [inline systemd/escape.go:83]: hm maybe we should just switch to using go-systemd? It's already an indirect dependency of ours. From the looks of it debian has v22 https://packages.debian.org/search?suite=all&searchon=names&keywords=go-systemd and so does Fedora https://koji.fedoraproject.org/koji/packageinfo?packageID=29480 and there's probably other places we could use it too

**Review by bboozzoo — DISMISSED** [review]
LGTM

**Review by pedronis — CHANGES_REQUESTED** [review]
I think securityTagFromCgroupPath in scanning.go needs changes to, as it goes back from the cgroup to a tag? do we need unescape support for that?

**Comment by andrewphelpsj** [comment]
@pedronis what do you think about swapping to using the escape/unescape functions from go-systemd, and getting rid of ours? That'd allow me to use the unescape implementation from there too.

I'm only hesitant because all of our tests don't pass when using that implementation:
```
phelps@ubuntu ~/s/e/systemd> go test

----------------------------------------------------------------------
FAIL: escape_test.go:28: SystemdTestSuite.TestEscapePath

escape_test.go:152:
    c.Check(unit.UnitNamePathEscape(t.in), Equals, t.out, Commentf(t.comment+" (with input %q)", t.in))
... obtained string = "top_level_1-..-top_level_2"
... expected string = "top_level_2"
... invalid path (with input "/top_level_1/../top_level_2/")

escape_test.go:152:
    c.Check(unit.UnitNamePathEscape(t.in), Equals, t.out, Commentf(t.comment+" (with input %q)", t.in))
... obtained string = "\\x9"
... expected string = "\\x09"
... tab character (with input "\t")

OOPS: 117 passed, 1 FAILED
--- FAIL: Test (0.56s)
FAIL
exit status 1
FAIL    github.com/snapcore/snapd/systemd       0.562s
```
I have no clue if these cases are relevant to our use case though. Especially that first case, which has some details in a comment here:
```
// TODO: this case shouldn't work it should cause an error to be
// fully compatible with systemd-escape(1), but it's not documented
// how or why systemd-escape considers this kind of path to be
// invalid, so for now we just process it in an expected and
// deterministic way
```

**Comment by bboozzoo** [comment]
```
escape_test.go:152:
    c.Check(unit.UnitNamePathEscape(t.in), Equals, t.out, Commentf(t.comment+" (with input %q)", t.in))
... obtained string = "top_level_1-..-top_level_2"
... expected string = "top_level_2"
... invalid path (with input "/top_level_1/../top_level_2/")
```

both implementations seem to be wrong:
```sh
$ systemd-escape -p top_level_1-..-top_level_2
Input 'top_level_1-..-top_level_2' is not an absolute file system path, escaping is likely not going to be reversible.
top_level_1\x2d..\x2dtop_level_2
```

escape_test.go:152:
    c.Check(unit.UnitNamePathEscape(t.in), Equals, t.out, Commentf(t.comment+" (with input %q)", t.in))
... obtained string = "\\x9"
... expected string = "\\x09"
... tab character (with input "\t")
```

yeah, this appears to be wrong, this is what I get:

```sh
$ systemd-escape -p "$(printf '\t')"
Input ' ' is not an absolute file system path, escaping is likely not going to be reversible.
\x09
```

systemd 255

I'd extend some test cases to cover valid snap mount paths and we should be good. We are already quite conservative with what is allowed in the snap/app names, this should be covered.

**Comment by andrewphelpsj** [comment]
@pedronis after some more reading and some talks with @bboozzoo, it looks like we can't quite use the unescape functions from go-systemd. The reason is that we use the `-` character in the unit names we create right now, which technically should be escaped if we intend it to be a `-`. `systemd-escape --unescape` (and the go-systemd implementation) interprets the `-` as an escaped `/`, we can't use that if we keep using `-` characters in the unit names.

To address this, I introduced two functions for converting between a security tag and a unit name. These functions only really deal with `+`, since we need to maintain the `-` characters, and all of the other characters that might appear in a security tag are already valid unit name characters.

File: systemd/escape.go [hunk 2]
```diff
@@ -67,3 +67,47 @@ func EscapeUnitNamePath(in string) string {
...
+// character in unit names. Note that these are still valid unit names, since
+// '-' is used by systemd-escape to represent the '/' character.
+//
+// To allow us to correctly convert between security tags and unit names (and to
+// maintain snapd's usage of '-' in unit names), this implementation only
+// escapes the '+' character, which was introduced with snap components.
+//
+// Examples of conversion:
+//   - "snap.name.app" -> "snap.name.app"
+//   - "snap.name+comp.hook.install" -> "snap.name\x2bcomp.hook.install"
```
Comment (bboozzoo) [inline systemd/escape.go:87]:
```suggestion
//   - "snap.name.app" -> "snap.name.app"
//   - "snap.some-name.some-app" -> "snap.some-name.some-app"
//   - "snap.name+comp.hook.install" -> "snap.name\x2bcomp.hook.install"
```

File: systemd/escape_test.go [hunk 3]
```diff
@@ -152,3 +152,57 @@ func (ts *SystemdTestSuite) TestEscapePath(c *C) {
...
+
+	for _, t := range cases {
+		unit, err := systemd.UnitNameFromSecurityTag(t.tag)
+		if t.err == "" {
+			c.Check(unit, Equals, t.unit)
+			c.Check(err, IsNil)
+
+			// if we don't expect the conversion to fail, check that the inverse
+			// function works too
+			tag := systemd.SecurityTagFromUnitName(unit)
```
Comment (bboozzoo) [inline systemd/escape_test.go:202]: Even though it'd be very simple right now a separate unit test for `SecurityTagFromUnitName` would be useful for future refactors.

**Review by pedronis — APPROVED** [review]
thank you, couple small comments

File: snap/naming/tag.go [hunk 4]
```diff
@@ -101,27 +111,46 @@ func ParseSecurityTag(tag string) (SecurityTag, error) {
...
-		return nil, errInvalidSecurityTag
-	}
 	// Depending on the type of the tag we either expect application name or
 	// the "hook" literal and the hook name.
 	if len(parts) == 3 {
+		// TODO: if components get apps, we can lift this check out of the if
+		// statement and move the strings.Cut to earlier in the function
+		if err := ValidateInstance(snapName); err != nil {
+			return nil, errInvalidSecurityTag
+		}
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply snap/naming/tag.go:121]: Good call, thank you.

**Comment by andrewphelpsj** [comment]
blocked until 2.63 is cut

**Pull request merged**